patches/server/1073-Extend-HumanEntity-dropItem-API.patch

```diff
@@ -0,0 +1,95 @@
+     }
+ 
++    // Paper start - Extend HumanEntity#dropItem API
++    @Override
++    public @org.jetbrains.annotations.Nullable org.bukkit.entity.Item dropItem(final ItemStack itemStack) {
++        return this.dropItem(itemStack, false, false);
++    }
++
++    @Override
++    public @org.jetbrains.annotations.Nullable org.bukkit.entity.Item dropItem(final ItemStack itemStack, final boolean retainOwnership) {
++        return this.dropItem(itemStack, retainOwnership, false);
++    }
++
++    @Override
++    public @org.jetbrains.annotations.Nullable org.bukkit.entity.Item dropItem(final ItemStack itemStack, final boolean retainOwnership, final boolean throwRandomly) {
++        return this.dropItem(this.inventory.first(itemStack), retainOwnership, throwRandomly);
++    }
++
++    @Override
++    public @org.jetbrains.annotations.Nullable org.bukkit.entity.Item dropItem(final int slot) {
++        return this.dropItem(slot, false, false);
++    }
++
++    @Override
++    public @org.jetbrains.annotations.Nullable org.bukkit.entity.Item dropItem(final int slot, final boolean retainOwnership) {
++        return this.dropItem(slot, retainOwnership, false);
++    }
++
++    @Override
++    public @org.jetbrains.annotations.Nullable org.bukkit.entity.Item dropItem(final int slot, final boolean retainOwnership, final boolean throwRandomly) {
++        // Make sure the slot is in bounds
++        if (slot < 0 || this.inventory.getSize() <= slot) {
++            return null;
++        }
++
++        final ItemStack stack = this.inventory.getItem(slot);
++        final org.bukkit.entity.Item itemEntity = dropItemRaw(stack, retainOwnership, throwRandomly);
++
++        this.inventory.setItem(slot, null);
++        return itemEntity;
++    }
++
++    @Override
++    public @org.jetbrains.annotations.Nullable org.bukkit.entity.Item dropItem(final org.bukkit.inventory.EquipmentSlot slot) {
++        return dropItem(slot, false, false);
++    }
++
++    @Override
++    public @org.jetbrains.annotations.Nullable org.bukkit.entity.Item dropItem(final org.bukkit.inventory.EquipmentSlot slot, final boolean retainOwnership) {
++        return dropItem(slot, retainOwnership, false);
++    }
++
++    @Override
++    public @org.jetbrains.annotations.Nullable org.bukkit.entity.Item dropItem(final org.bukkit.inventory.EquipmentSlot slot, final boolean retainOwnership, final boolean throwRandomly) {
++        final ItemStack stack = this.inventory.getItem(slot);
++        final org.bukkit.entity.Item itemEntity = dropItemRaw(stack, retainOwnership, throwRandomly);
++
++        this.inventory.setItem(slot, null);
++        return itemEntity;
++    }
++
++    private org.bukkit.entity.Item dropItemRaw(final ItemStack is, final boolean retainOwnership, final boolean throwRandomly) {
++        if (!(this.getHandle() instanceof ServerPlayer player)) {
++            return null;
++        }
++
++        if (is == null || is.getType() == Material.AIR) {
```

Comment:
Probably an exception would be more appropriate

Reply:
I would rather make it return null, as in no item was found, so no item was dropped. Because otherwise you'd have to manually run a check for whether the item actually exists, and it just turns into a mess

Reply:
valid point